File: core/src/main/java/heroes/journey/entities/actions/TargetAction.java
package heroes.journey.entities.actions;

import static heroes.journey.registries.Registries.ActionManager;

import java.util.ArrayList;
import java.util.List;

import heroes.journey.entities.actions.inputs.ActionInput;
import heroes.journey.entities.actions.inputs.TargetInput;
import heroes.journey.entities.actions.results.AIOnSelectNotFound;
import heroes.journey.entities.actions.results.ActionListResult;
import heroes.journey.entities.actions.results.ActionResult;

public abstract class TargetAction<I> extends Action {

  public TargetAction(String id, String name, String description, Cost cost) {
    super(id, name, description, true, cost);
  }

  public abstract List<I> getTargets(ActionInput input);

  public String getTargetDisplayName(TargetInput<I> input) {
    return input.getInput().toString();
  }

  @Override
  public ShowAction requirementsMet(ActionInput input) {
    List<I> options = getTargets(input);
    if (options.isEmpty())
      return ShowAction.GRAYED;
    return super.requirementsMet(input);
  }

  @Override
  public ActionResult internalOnSelect(ActionInput input) {
    List<Action> actionOptions = new ArrayList<>();
    List<I> options = getTargets(input);
    for (I option : options) {
      TargetInput<I> targetInput = new TargetInput<>(input.getGameState(), input.getEntityId(), option);
      actionOptions.add(getAction(targetInput));
    }
    return new ActionListResult(actionOptions);
  }

  public Action getActionFromSelected(ActionInput inputBase, String selectedOption) {
    List<I> options = getTargets(inputBase);
    for (I option : options) {
      if (option.toString().equals(selectedOption)) {
        TargetInput<I> targetInput = new TargetInput<>(inputBase.getGameState(), inputBase.getEntityId(),
            option);
        cost.onUse(targetInput);
        return getAction(targetInput);
      }
    }
    throw new RuntimeException("Selected Target option not found " + selectedOption);
  }

  private Action getAction(TargetInput<I> targetInput) {
    String displayName = getTargetDisplayName(targetInput);
    I option = targetInput.getInput();
    return new Action(getId() + "," + option.toString(), displayName, "", false, null) {

      public ShowAction internalRequirementsMet(ActionInput input) {
        return requirementsMetTarget(targetInput);
      }

      public void internalOnHover(ActionInput input) {
        onHoverTarget(targetInput);
      }

      public ActionResult internalOnSelect(ActionInput input) {
        return onSelectTarget(targetInput);
      }

      public ActionResult internalOnSelectAI(ActionInput input) {
        return onSelectAITarget(targetInput);
      }
    };
  }

  protected ShowAction requirementsMetTarget(TargetInput<I> targetInput) {
    return ShowAction.YES;
  }

  protected ActionResult onSelectAITarget(TargetInput<I> targetInput) {
    return new AIOnSelectNotFound();
  }

  protected abstract ActionResult onSelectTarget(TargetInput<I> targetInput);

  protected void onHoverTarget(TargetInput<I> targetInput) {
  }

  public TargetAction<I> register() {
    ActionManager.register(this);
    return this;
  }

}

File: core/src/main/java/heroes/journey/entities/actions/TargetAction.kt
package heroes.journey.entities.actions

import heroes.journey.modlib.actions.ActionEntry
import heroes.journey.modlib.actions.ShowAction
import heroes.journey.modlib.actions.results.AIOnSelectNotFound
import heroes.journey.modlib.actions.results.ActionListResult
import heroes.journey.modlib.actions.results.ActionResult
import heroes.journey.modlib.actions.results.NullResult
import heroes.journey.registries.Registries

class TargetAction<I>(
    id: String,
    requirementsMetFn: (ActionInput) -> ShowAction = { ShowAction.YES },
    onHoverFn: (ActionInput) -> Unit = {},
    inputDisplayNameFn: ((Map<String, String>) -> String)? = null,
    val getTargets: (ActionInput) -> List<I>,
    private val targetAction: String
) : Action(
    id,
    true,
    requirementsMetFn,
    onHoverFn,
    { NullResult() },
    { AIOnSelectNotFound() },
    inputDisplayNameFn
) {
    override fun requirementsMet(input: ActionInput): ShowAction {
        var targetShowAction: ShowAction = ShowAction.YES
        if (getTargets(input).isEmpty()) targetShowAction = ShowAction.GRAYED
        return super.requirementsMet(input).and(targetShowAction)
    }

    override fun onSelect(input: ActionInput, ai: Boolean): ActionResult {
        val actionOptions: List<ActionEntry> = getTargets(input).map { option ->
            val copyInput: HashMap<String, String> = input.getHashMapCopy()
            copyInput["target"] = option.toString()
            ActionEntry(targetAction, copyInput)
        }
        return ActionListResult(actionOptions)
    }

    override fun register(): TargetAction<I> {
        Registries.ActionManager.register(this)
        return this
    }
}
